### PR TITLE
Add `CloseSocketFunction` and export Curl constructor 

### PR DESCRIPTION
**Pull request by crtschin**
This makes the following changes:
- Add the `CURLOPT_CLOSESOCKETFUNCTION` curl opt
- Disambiguates the writer function from the writer function used for headers in the marshalling debug messages
- Export the Curl data constructor and record fields

The first two are intended to help with debugging. The last one is so the user can call `curl_easy_cleanup` without having to wait on the garbage collector to clean up the handle.

### CODE REVIEW

File: Network/Curl/Opts.hs [hunk 1]
```diff
@@ -281,6 +282,11 @@ type OpenSocketFunction
   -> Ptr () -- ^ The struct describing the curl_sockaddr struct
   -> IO CInt -- ^ The opened socket or CURL_SOCKET_BAD
 
+type CloseSocketFunction
+   = Ptr ()  -- ^ Pointer to clientp data set by OpenSocketData
+  -> CInt    -- ^ The curl socket type
+  -> IO CInt -- ^ Whether closing was successful
```
Comment (joris-burgers) [inline Network/Curl/Opts.hs]: What are the values when closing is successful, 0 and 1?

File: Network/Curl/Easy.hs [hunk 2]
```diff
@@ -112,6 +112,12 @@ setopt hh o = curlPrim hh $ \ r h -> unmarshallOption (easy_um r h) o
             fp <- mkWriter x
             updateCleanup r i $ debug "FREE: WRITER" >> freeHaskellFunPtr fp
             liftM toCode $ easy_setopt_wfun h i fp
+     , u_writeFunH -- :: Int -> WriteFunction -> IO a
+       = \ i x -> do
+            debug "ALLOC: HEADER WRITER"
+            fp <- mkWriter x
+            updateCleanup r i $ debug "FREE: HEADER WRITER" >> freeHaskellFunPtr fp
```
Comment (joris-burgers) [inline Network/Curl/Easy.hs:119]: We use the `freeHaskellFunPtf fp`, but still use `fp` after that. Is that safe?

File: Network/Curl/Opts.hs [hunk 3]
```diff
@@ -539,6 +548,7 @@ verboseUnmarshaller u =
     , u_bytestring  = twoS "u_bytestring" u_bytestring
     , u_ptr         = twoS "u_ptr" u_ptr
     , u_writeFun    = two "u_writeFun" u_writeFun
+    , u_writeFunH   = two "u_writeFun" u_writeFun
```
Comment (joris-burgers) [inline Network/Curl/Opts.hs]: Shouldn't this be `two "u_writeFunH" u_writeFunH`?

**Review by joris-burgers — COMMENTED** [review]
I am no expert in this code, so I commented about some parts that were counterintuitive to me, but they might be correct :man_shrugging: .

File: Network/Curl/Opts.hs (hunk 1 above)
Comment (crtschin) [reply Network/Curl/Opts.hs]: Precisely, in typical c style 0 if no error else 1.

File: Network/Curl/Easy.hs (hunk 2 above)
Comment (crtschin) [reply Network/Curl/Easy.hs:119]: I don't quite understand, where do we use `fp` after `freeHaskellFunPtf fp`? The cleanup functions are ran by a corresponding `runCleanup` call in either the finalizer of the `Curl` handle or a call to `reset`. Both of which are preceded by respectively invalidating the handle or resetting to its initial state,.

Comment (joris-burgers) [reply Network/Curl/Easy.hs:119]: In the final line, we have `liftM toCode $ easy_setopt_wfun h i fp`, so we use `fp` after that. On the other hand, all other places use it as well, so it is probably fine.

File: Network/Curl/Opts.hs (hunk 1 above)
Comment (joris-burgers) [reply Network/Curl/Opts.hs]: Could you maybe document that, not everyone might be familiar with that convention.

File: Network/Curl/Opts.hs [hunk 4]
```diff
@@ -539,6 +548,7 @@ verboseUnmarshaller u =
     , u_bytestring  = twoS "u_bytestring" u_bytestring
     , u_ptr         = twoS "u_ptr" u_ptr
     , u_writeFun    = two "u_writeFun" u_writeFun
+    , u_writeFunH   = two "u_writeFunH" u_writeFun
```
Comment (joris-burgers) [inline Network/Curl/Opts.hs]: Shouldn't they both not be `u_writeFunH`, not only the first one?

File: Network/Curl/Easy.hs (hunk 2 above)
Comment (crtschin) [reply Network/Curl/Easy.hs:119]: Yeah, the existing code is pretty obscurely written in that `updateCleanup` adds an `IO ()` action to be run when `runCleanup` is called. So the `fp` is still allocated when `liftM` happens but is deallocated when the `Curl` handle is garbage collected or `reset`.

Comment (joris-burgers) [reply Network/Curl/Easy.hs:119]: Ah, that makes sense, thanks for the explanation.

**Review by joris-burgers — APPROVED** [review]
LGTM!

**Comment by crtschin** [comment]
@OpsBotPrime merge

**Pull request merged**